src/pool-bin/libraries/BinHooks.sol

```diff
@@ -0,0 +1,157 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+// Copyright (C) 2024 PancakeSwap
+pragma solidity ^0.8.24;
+
+import "../interfaces/IBinHooks.sol";
+import {PoolKey} from "../../types/PoolKey.sol";
+import {IBinPoolManager} from "../interfaces/IBinPoolManager.sol";
+import {Hooks} from "../../libraries/Hooks.sol";
+import {BalanceDelta, BalanceDeltaLibrary, toBalanceDelta} from "../../types/BalanceDelta.sol";
+
+library BinHooks {
+    using Hooks for bytes32;
+
+    function validate(PoolKey memory key) internal pure {
+        if (
+            key.parameters.hasOffsetEnabled(HOOKS_BEFORE_SWAP_RETURNS_DELTA_OFFSET)
+                && !key.parameters.hasOffsetEnabled(HOOKS_BEFORE_SWAP_OFFSET)
+        ) {
+            revert Hooks.HookPermissionsValidationError();
+        }
+
+        if (
+            key.parameters.hasOffsetEnabled(HOOKS_AFTER_SWAP_RETURNS_DELTA_OFFSET)
+                && !key.parameters.hasOffsetEnabled(HOOKS_AFTER_SWAP_OFFSET)
+        ) {
+            revert Hooks.HookPermissionsValidationError();
+        }
+
+        if (
+            key.parameters.hasOffsetEnabled(HOOKS_AFTER_MINT_RETURNS_DELTA_OFFSET)
+                && !key.parameters.hasOffsetEnabled(HOOKS_AFTER_MINT_OFFSET)
+        ) {
+            revert Hooks.HookPermissionsValidationError();
+        }
+
+        if (
+            key.parameters.hasOffsetEnabled(HOOKS_AFTER_BURN_RETURNS_DELTA_OFFSET)
+                && !key.parameters.hasOffsetEnabled(HOOKS_AFTER_BURN_OFFSET)
+        ) {
+            revert Hooks.HookPermissionsValidationError();
+        }
+    }
+
+    function afterMint(
+        PoolKey memory key,
+        IBinPoolManager.MintParams memory params,
+        BalanceDelta delta,
+        bytes calldata hookData
+    ) internal returns (BalanceDelta callerDelta, BalanceDelta hookDelta) {
+        IBinHooks hooks = IBinHooks(address(key.hooks));
+        callerDelta = delta;
```

Comment:
if skip i.e. caller is already hook, then `callerDelta` is unchanged and hookDelta is 0